Run ceph-fuse in isolated systemd scope

cofyc opened the pull request:
**What this PR does / why we need it**:

Run ceph-fuse in isolated systemd scope. Then on kubelet restart, fuse daemon will not be killed.

Why we need to run fuse daemon in its own system scope, see https://github.com/kubernetes/kubernetes/pull/49640.
 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #67643

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fix cephfs plugin to mount cephfs fuse in isolated systemd scope to prevent ceph fuse daemons from being killed on kubelet restart.
```

cofyc commented:
/assign @jsafrane 
cc @rootfs @zhangxiaoyu-zidif 
cc @choury seems like you are using cephfs in large cluster

cofyc commented:
Found [mount.ceph.fuse](https://github.com/ceph/ceph/blob/master/src/mount.fuse.ceph#L75) does not exit non-zero on failure.

choury commented:
Jewel's [mount.fuse.ceph](https://github.com/ceph/ceph/blob/jewel/src/mount.fuse.ceph) only support old format.

Commit:
Run ceph-fuse in isolated systemd scope.

cofyc commented:
I've updated PR to run `ceph-fuse` command in isolated systemd scope.
@jsafrane Please take a look at. By the way, I moved systemd utilites to a separate package.

jsafrane commented on pkg/volume/cephfs/cephfs.go:
```diff
@@ -419,11 +418,26 @@ func (cephfsVolume *cephfs) execFuseMount(mountpoint string) error {
+	output, err := exec.Run(command, mountArgs...)
 	if err != nil || !(strings.Contains(string(output), "starting fuse")) {
-		return fmt.Errorf("Ceph-fuse failed: %v\narguments: %s\nOutput: %s\n", err, mountArgs, string(output))
+		return fmt.Errorf("Mouting ceph-fuse failed: %v\ncommand: %q\narguments: %s\noutput: %s\n", err, command, mountArgs, string(output))
```
typo: mouting

cofyc replied:
fixed

cofyc commented:
This PR is a quick fix. I'm not sure if it's necessary to add it in mounter.Interface. Some alternative designs:

```
command, args = plugin.mounter.AddSystemdScope(command, args)
output, err := plugin.exec.Run(command, args)
```

or 

```
output, err := plugin.mounter.RunInSystemdScope(command, args) # use mounter internal exec 
```

or a helper function

```
output, err := mount.RunInSystemdScope(exec, command, args) # like current design, but only export one function
```

msau42 commented:
I'd prefer not leaking Linux into the interface.  Can we somehow encapsulate this functionality into something more general?  Also, will CSI run into the same issue?  We're trying to refactor the Mount interface so that CSI plugins can import common mounting libraries without having to pull in all of kubernetes.

cofyc commented:
> Also, will CSI run into the same issue? 

It depends on how CSI plugin driver is deployed.

If processes run in a shared cgroup (e.g. root cgroup), there is no problem.
If processes run in systemd or docker containers, there will be a dedicated control group for them, all its forked processes (including daemonized ones, e.g. fuse daemons) will be run in this cgroup by default. On systemd restart or container restart, processes in this cgroup will be killed.

To prevent from being killed, we need to run forked processes in a separate cgroup.

For systemd, if process is launched by `systemd-run --scope`, process will be run in its own scope, e.g.

```
$ sudo systemd-run --scope bash -c 'cat /proc/$$/cgroup'
Running scope as unit run-raf9d717ba644455d90d8ecd2ad265aa3.scope.
11:cpuset:/
10:hugetlb:/
9:pids:/system.slice/run-raf9d717ba644455d90d8ecd2ad265aa3.scope
8:cpu,cpuacct:/system.slice/run-raf9d717ba644455d90d8ecd2ad265aa3.scope
7:devices:/system.slice/run-raf9d717ba644455d90d8ecd2ad265aa3.scope
6:perf_event:/
5:memory:/system.slice/run-raf9d717ba644455d90d8ecd2ad265aa3.scope
4:blkio:/system.slice/run-raf9d717ba644455d90d8ecd2ad265aa3.scope
3:net_cls,net_prio:/
2:freezer:/
1:name=systemd:/system.slice/run-raf9d717ba644455d90d8ecd2ad265aa3.scope
```

For docker containers or kubernetes pods, there are two ways to escape current cgroup:

1) Using nsenter to run fuse daemons in host namespace, as @jsafrane said, this requires fuse daemons on the host.

```
$ docker run -it --rm -v /:/rootfs --privileged alpine sh
/ # nsenter --mount=/rootfs/proc/1/ns/mnt -- bash -c 'cat /proc/$$/cgroup'
11:cpuset:/
10:hugetlb:/
9:pids:/
8:cpu,cpuacct:/
7:devices:/
6:perf_event:/
5:memory:/
4:blkio:/
3:net_cls,net_prio:/
2:freezer:/
1:name=systemd:/
```

2) Run fuse daemons in separate docker containers or pods. This is like `systemd-run --scope`.

> I'd prefer not leaking Linux into the interface. Can we somehow encapsulate this functionality into something more general? 

Sometimes, users may need to run fuse daemons in its own cgroup, so they can limit resources etc. Perhaps it's better to be configurable. I'll update when I have some ideas.

cofyc commented:
/close
not urgent

warmchang commented:
/cc
